### PR TITLE
Reduce allocations in go driver for time conversions

### PR DESCRIPTION
**Pull request by derekperkins**
Admittedly cargo culting this change to reduce go allocations. The original author is @achille-roussel as a PR here: https://github.com/go-sql-driver/mysql/pull/615

More detail behind the change at https://segment.com/blog/allocation-efficiency-in-high-performance-go-services/

We could probably also move the `t.Location()` check inside the else statement here for zero time.

### CODE REVIEW

File: go/vt/vitessdriver/time.go [hunk 1]
```diff
@@ -109,5 +109,16 @@ func NewDatetime(t time.Time, defaultLoc *time.Location) sqltypes.Value {
 	if t.Location() != defaultLoc {
 		t = t.In(defaultLoc)
 	}
-	return sqltypes.MakeTrusted(sqltypes.Datetime, []byte(t.Format(isoTimeFormat)))
+
+	// using this backing array and AppendFormat reduces heap allocations
+	var a [64]byte
+	var b = a[:0]
```
Comment (sougou) [inline go/vt/vitessdriver/time.go]: Good catch. Since this variable is escaping, it's better to use `b := make([]byte, 0, 64)`

Comment (derekperkins) [reply go/vt/vitessdriver/time.go]: I can't take any credit other than that I read articles by smart people. :) I just pushed a commit with your suggested change.

**Comment by derekperkins** [comment]
I went ahead and moved the location check

**Comment by sougou** [comment]
I'll let @vmg take a second look.

**Comment by vmg** [comment]
Very interesting PR! Thanks for CCing me @sougou!

The original optimization in `go-sql-driver/mysql` is veeery nice because it tweaks the flow of the Time formatting logic to make it truly zero-allocation: The MySQL driver is attempting to build a packet buffer with the contents it needs to send to MySQL, and as part of building that buffer, it is serializing `time.Time` data directly into it.

Here's the logic in the `mysql` driver after the optimization:

```go
var a [64]byte
var b = a[:0]

if v.IsZero() {
	b = append(b, "0000-00-00"...)
} else {
	b = v.In(mc.cfg.Loc).AppendFormat(b, timeFormat)
}
paramValues = appendLengthEncodedInteger(paramValues,
	uint64(len(b)),
)
paramValues = append(paramValues, b...)
```

Here, `paramValues` is the packet buffer, which is already pre-allocated to a reasonable size.  By using the `AppendFormat` method on `Time` to serialize the time into an slice which is backed by a _stack allocated array_, it is then possible to append that slice directly into `paramValues`. (note that the intermediate stack-allocated array is necessary because we need to serialize the _length_ of `b` to `paramValues` before we can serialize `b` itself -- otherwise we could just `AppendFormat` directly into `paramValues`, which would be an even more elegant optimization!)

Either way, the result is serialization logic with zero-allocation cost, because the compiler can track that neither `a` or `b` escape the current scope. The result of the serialization does get written to `paramValues`, but that slice was already heap-allocated, so nothing is lost.

The case in the Vitess driver, however, is a bit more complicated: In `NewDateTime`, we're not serializing to an existing buffer; we're creating a new `sqltypes.Value`, which will contain the serialized bytes of our `Time` inside:

```go
package sqltypes

type Value struct {
	typ querypb.Type
	val []byte
}

func MakeTrusted(typ querypb.Type, val []byte) Value {
	if typ == Null {
		return NULL
	}
	return Value{typ: typ, val: val}
}
```

Here we can tell that the bytes stored in `Value.val` _must_ escape to the heap, because it's impossible to track where we're storing the `Value` that `NewDateTime()` returns.

It looks like @derekperkins was onto something here, because this PR has no stack-backed array. `b` is declared _directly on the heap_ with a `make()`, so by definition it cannot be zero-allocation:

```go
func NewDatetime(t time.Time, defaultLoc *time.Location) sqltypes.Value {
	// using this backing array and AppendFormat reduces heap allocations
	var b = make([]byte, 0, 64)

	if t.IsZero() {
		b = append(b, "0000-00-00"...)
	} else {
		if t.Location() != defaultLoc {
			t = t.In(defaultLoc)
		}
		b = t.AppendFormat(b, isoTimeFormat)
	}

	return sqltypes.MakeTrusted(sqltypes.Datetime, b)
}
```

Is this a missed optimization, then? No, it is not, as we're storing `b` in `Value.val`, which means it will need to be heap-allocated anyway. If Derek had written `b` as a stack-backed slice (`var a byte[64]; var b = a[:0]`), the compiler would have escaped `a` to the heap all the same, and we'd be in the same situation.

So does that mean then that this optimization is a no-op? Not quite. Before this PR, we were calling `Time.Format` and converting its `string` return value into `[]byte`. In theory, this conversion requires an allocation, because `string`s in Go are immutable, and `[]byte` slices are not...

But wait, the return of `Time.Format` is an r-value! Will the Go compiler be smart enough to optimize away the `string` -> `[]byte` conversion, considering that it can prove that the original `string` will never be referenced again? Let's find out:

```go

type Value struct {
	v []byte
}

func TimeNaive(t time.Time) Value {
	return Value{v: []byte(t.Format(isoTimeFormat))}
}

func TimeOptimized(t time.Time) Value {
	var b = make([]byte, 0, 64)
	b = t.AppendFormat(b, isoTimeFormat)
	return Value{v: b}
}

func BenchmarkTimeFormat(b *testing.B) {
	t := time.Now()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		TimeNaive(t)
	}
}

func BenchmarkTimeOptimized(b *testing.B) {
	t := time.Now()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		TimeOptimized(t)
	}
}
```

```
$ go test -bench=.
goos: linux
goarch: amd64
BenchmarkTimeFormat-6      	 5000000	       273 ns/op	      64 B/op	       2 allocs/op
BenchmarkTimeOptimized-6   	 5000000	       254 ns/op	      64 B/op	       1 allocs/op
```

Check that out! It really looks like the Go compiler can optimize away this conversion: even though the "old code" shows 2 allocations per loop, both implementations are only allocating 64 bytes on the heap! The second allocation in the old code seems to be a no-op.

In conclusion: very good idea for a PR, but it appears that the Go compiler already has our back here, and the resulting code is already as optimal as it can be.

What a neat little optimization problem. Now I wonder: did I miss something with my benchmarks? Was my approach correct? Is it sensible to depend on this compiler optimization? Should we merge this PR all the same, as it does shave a couple `ns` per run?

I'm leaning towards "the current code is simpler and hence fine as it is", but I'd love to hear your thoughts.

**Comment by derekperkins** [comment]
@vmg 👏  👏  👏   That's the best PR review I've ever seen! Like I said, this was a fairly naive attempt to optimize in the same way that it worked in `go-sql-driver`. Given your benchmarking and analysis, I'll go ahead and close this PR. I think we can be confident relying on the compiler here.

**Comment by sougou** [comment]
This has been a lingering question in my mind. Thanks for the research :).
